### DISCUSSION TITLE
Dynamically change and map data from getServerSideProps() when the dropdown is used

### DISCUSSION THREAD

**Discussion by Kvaibhav01** [vercel/next.js#42628] [Help] [answered]
Hello!

I want to implement a language switcher where if I change the language value from the dropdown, it should fetch the data from the correct locale API path. My translation files are in an Express project with these endpoints:

`http://localhost:3001/en` and `http://localhost:3001/es` for `en` and es locales, respectively. I want, by default only `en` is fetched so I did this in getServerSideProps()`:

```js
export async function getServerSideProps() {
  const data = fetch('http://localhost:3001/en');
  const fetchedData = await (await data).json();

  return {
    props: {
      fetchedData,
    },
  };
}
```

And I can correctly render it like:

```jsx
export default function Home({ fetchedData }) {
  const router = useRouter();

  return (
    <div>
      //...

      {/* Using dropdown to push the route to correctly chosen locale value */}
      <select onChange={(e) => router.push(e.target.value)}>
        <option value='en'>English</option>
        <option value='es'>Español</option>
      </select>

      <div>
        {fetchedData.map((news, id) => (
          <div
            key={id}
          >
            <p>
              {news.date}
            </p>
           //...
          </div>
        ))}
      </div>
    </div>
  );
}
```

Now what I want is that along with the change in route, when we select a different locale value (like `es`) from dropdown select, it fetches from `http://localhost:3001/es` and then maps over in the `Home` component, making it dynamic.

How can I achieve this?

I passed `context` to `getServerSideProps()`, but how can I pull `es` and `en` from there?

**Comment by icyJoseph** [comment]
Can you share how's your page folder structured?

I suspect that the same page handles English and Spanish? Is home `/`? 

In short, how are you thinking of doing this? With just one page `page/index.tsx`?

**Comment by Kvaibhav01** [comment]
Here's how the _pages_ folder is structured:

<img width="378" alt="A screenshot of page folder structure" src="https://user-images.githubusercontent.com/11731837/200580802-d7714c59-e372-4fab-9eb3-d21933633d95.png">


> I suspect that the same page handles English and Spanish? Is home /?

Yes, **the _pages/index.js_ file is where I am trying to render the UI** with data coming from an Express local API (`http://localhost:3001/en` and `http://localhost:3001/es`). When you run the app, it goes to `http://localhost:3001` and when we want to change language, it redirects to the respective route like `http://localhost:3001/es` for the `es` value in `select` and `http://localhost:3001/en` for `en`.

> In short, how are you thinking of doing this? With just one page page/index.tsx?

Yes, on the homepage, I want to render all the data items I am getting from the Express backend. It lists news data like news title, a description, date of publish etc. **By default I want to render `en` locale but in the navbar dropdown I want the user can change the language to Spanish and when they do that, they should be redirected to `es` locale**. Here, just when the locale is changed, I want to fetch data from `http://localhost:3001/es`.

**Reply by icyJoseph** [reply]
Alright, gimme 5 minutes. I am assuming you have a next.config.js `i18n` object setup.

**Reply by Kvaibhav01** [reply]
Yes, it is setup like this:

```js
const { i18n } = require('./next-i18next.config.js');

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false,
  swcMinify: true,
  webpack: (config) => {
    config.resolve.fallback = { fs: false };
    return config;
  },
  i18n,
};

module.exports = nextConfig;
```

And I am using the [next-i18next](https://github.com/i18next/next-i18next) package to localize static elements like the button label that says "Read more", webpage heading, dropdown label, etc. But I want the dynamic content, which is the data coming from an Express server `http://localhost:3001/en`, to be fetched on demand only when the language is changed.

**Comment by icyJoseph** [comment, marked as answer]
Try with:

```js
import { useRouter } from "next/router";
import { useState } from "react";
import styles from "../styles/Home.module.css";

export default function Home({ message }) {
  const router = useRouter();

  const [counter, setCounter] = useState(0);

  return (
    <div className={styles.container}>
      {message}
      <select
        onChange={(e) =>
          router.push(router.pathname, router.pathname, {
            locale: e.target.value,
          })
        }
      >
        <option value="en">English</option>
        <option value="es">Español</option>
      </select>

      <button onClick={() => setCounter((x) => x + 1)}>{counter} + 1</button>
    </div>
  );
}

// probably need to make this `async (context) => {`
export const getServerSideProps = (context) => {
  const { locale } = context;
  const message = locale === "en" ? "Hello World" : "Hola Mundo";
  return { props: { message } };
};
```

And in your `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  i18n: {
    locales: ["en", "es"],
    // This is the default locale you want to be used when visiting
    // a non-locale prefixed path e.g. `/hello`
    defaultLocale: "en",
  },
};

module.exports = nextConfig;
```

I put a counter, so you can see that between locale transitions the counter state is preserved.

**Reply by Kvaibhav01** [reply]
Thanks for the code example!

The transition from `en` to `es` for static content (which is not fetched by an external API call), like the `message` is working well. That's not an issue. What we need is that when the `locale` is changed from `en` to `es` or vice versa, it fetches data from the correct API call (`http://localhost:3001/en` or `http://localhost:3001/es`) and renders that specific data in `getServerSideProps`.

How can we do that? I assume passing a `context` to `getServerSideProps` can help, but how can we pass the `context` for the `locale` change for `en` and `es`?

**Reply by icyJoseph** [reply]
Yeah, but you can just read the locale from the context, because when you do the `router.push(router.pathname, router.pathname, {locale: locale})`, it'll refetch `getServerSideProps`. The current locale is part of the `GSSP` context.

```js
export const getServerSideProps = async (context) => {
  const { locale } = context; // -> Here!
  const message = locale === "en" ? "Hello World" : "Hola Mundo";
  const res = await fetch('http://localhost:3001/' + locale);
  const data = await res.json();
  return { props: { message, data } };
};
```

**Reply by icyJoseph** [reply]
Here's a post which can also help you out, https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/, the post is exclusively about triggering new props from getServerSideProps, without losing page state. In the post they use `.replace` instead of `.push`.

**Reply by Kvaibhav01** [reply]
Hey @icyJoseph thanks for your help. It works! 🎉 

I didn't knew about this:

> when you do the `router.push(router.pathname, router.pathname, {locale: locale})`, it'll refetch `getServerSideProps`. 

Thanks for letting me know that. And I will check that article too. 😊
